src/python/pants/backend/python/subsystems/pex_bootstrap/_ptex_launcher.py

```diff
@@ -0,0 +1,75 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+# This bootstrapping technique for lazily loading requirements comes from the pex project test case:
+# https://github.com/pantsbuild/pex/commit/66aa5ef69e8b65004a548b44edc0b81dfa8ec736#diff-b5b57c9d4fc663f59ecaedb851c13bd8R23
+
+import json
+import os
+import sys
+
+from pex import resolver
+from pex.common import open_zip
+from pex.fetcher import Fetcher, PyPIFetcher
+from pex.interpreter import PythonInterpreter
+from pex.pex_builder import PEXBuilder
+from pex.pex_info import PexInfo
+from pex.platforms import Platform
+
+
+self = sys.argv[0]
+filename_base, ext = tuple(os.path.splitext(self))
+hydrated_ipex_file = '{}.ipex'.format(filename_base)
+
+if not os.path.isfile(hydrated_ipex_file):
```

Comment:
I lost a longish comment!

> How expensive would it be to just unconditionally do the ipex creation?

Not at all!

> I can imagine this being a confusing caching experience if someone edits a python file, re-builds the file, and doesn't see the changes they expect to take effect.

I ran into exactly this frustration when developing this!

> I'd also be tempted to hydrate this into a tempdir or similar

Great idea! This addresses the use case where a redeployment just means one-shot executing a persistent server process (for example). The one extension I'm thinking of to cover the case where some tool which uses pexes built from pants expects to receive a `.ptex` file, and wants to hydrate it to a `.ipex` before possibly invoking it multiple times.

To address that, I'm thinking that extending `_ptex_launcher.py` to check for the env var `PANTS_PTEX_CREATE_IPEX=<filename>.ipex`, and if this is provided, to hydrate the file (unconditionally) to `<filename>.ipex`.

Reply:
> I lost a longish comment!
> 
> > How expensive would it be to just unconditionally do the ipex creation?
> 
> Not at all!
> 
> > I can imagine this being a confusing caching experience if someone edits a python file, re-builds the file, and doesn't see the changes they expect to take effect.
> 
> I ran into exactly this frustration when developing this!
> 
> > I'd also be tempted to hydrate this into a tempdir or similar
> 
> Great idea! This addresses the use case where a redeployment just means one-shot executing a persistent server process (for example). The one extension I'm thinking of to cover the case where some tool which uses pexes built from pants expects to receive a `.ptex` file, and wants to hydrate it to a `.ipex` before possibly invoking it multiple times.
> 
> To address that, I'm thinking that extending `_ptex_launcher.py` to check for the env var `PANTS_PTEX_CREATE_IPEX=<filename>.ipex`, and if this is provided, to hydrate the file (unconditionally) to `<filename>.ipex`.

The reason I asked how expensive it is to just make it each time is that if it's sufficiently cheap, we just don't need that option, people can just run this file and be happy. Are you saying the overhead of hydrating the ipex is sufficient that they'd have good reason to want to use that option? (And that they'd need to understand things like "this file isn't portable across machines", in order to correctly use it)

Reply:
```bash
> dmcclanahan@tw-mbp-dmcclanahan: ~/tools/pants-v3 18:06:58
X> ls idk
total 8.2k
-rw-r--r-- 1 dmcclanahan staff 107 Dec 13 18:01 BUILD
-rw-r--r-- 1 dmcclanahan staff  68 Dec 13 18:01 main.py
dmcclanahan@tw-mbp-dmcclanahan: ~/tools/pants-v3 18:07:22
X> cat idk/BUILD
python_binary(
  name='bin',
  dependencies=['examples/3rdparty/python:tensorflow'],
  source='main.py',
)
dmcclanahan@tw-mbp-dmcclanahan: ~/tools/pants-v3 18:07:26
X> cat idk/main.py
import tensorflow as tf

print(f'tf version was: {tf.__version__}')
dmcclanahan@tw-mbp-dmcclanahan: ~/tools/pants-v3 18:07:27
X> ./pants binary.py --generate-ptex --cache-ignore idk:bin
...
dmcclanahan@tw-mbp-dmcclanahan: ~/tools/pants-v3 18:07:40
X> time PANTS_PTEX_HYDRATE_ONLY_TO_IPEX=dist/bin.ipex ./dist/bin.ptex
Hydrating ./dist/bin.ptex to dist/bin.ipex...
PANTS_PTEX_HYDRATE_ONLY_TO_IPEX=dist/bin.ipex ./dist/bin.ptex  18.12s user 7.52s system 98% cpu 26.084 total
```

A 26 second startup time seems pretty significant, so I have instead annotated where PANTS_PTEX_HYDRATE_ONLY_TO_IPEX is consumed, otherwise just creating a temporary ipex and executing it.

Reply:
What's it _doing_ for 26 seconds? It should just be copying a handful of tiny files from one zip file to another, and generating one additional small text file? With any reasonable disk, that should be at most a second or so?

Is the extra time the time spent doing the resolve to populate the cache? What does a second run to a different path look like?

Reply:
Right now, when we call `.add_distribution()` (as per https://github.com/pantsbuild/pants/pull/8793#discussion_r357153474), we physically add these massive files to the `.ipex`, creating a very large file, with all the deps hydrated. As you described there, this should not need to happen.

One last thing @jsirois did in `test_issues_789_demo()` that I forgot was the use of `.pex_root`. I'm not sure if that's necessary here -- will look.

Reply:
Oh, we do set `cache=ipex_info.pex_root`, which I think was the trick. Will dig into where pex tries to resolve distributions from at runtime.